File: Data Structures & Algorithms/binary-search/submission-1.java
class Solution {
    public int search(int[] nums, int target) {
        return searchRecursive(nums, target, 0, nums.length - 1);
    }

    public int searchRecursive(int[] nums, int target, int left, int right) {
        int mid = left + (right - left) / 2;
        if (left > right) {
            return -1;
        }

        if (target == nums[mid]) {
            return mid;
        } else if (nums[mid] > target) {
            return searchRecursive(nums, target, left, mid - 1);
        } else {
            return searchRecursive(nums, target, mid + 1, right);
        }
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-1.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        numOfIslands = 0

        # full grid of -1 to track where I have visited to not repeat visit islands

        hasVisited = [["-1" for x in range(len(grid[0]))] for y in range(len(grid))]

        for y in range(len(grid)):
            for x in range(len(grid[0])):
                # We have not visited there yet
                if hasVisited[y][x] == "-1":
                    # We see an island
                    if grid[y][x] == "1":
                        numOfIslands += 1
                        # DFS or BFS to mark down entire island as visited
                        self.searchEntireIsland(grid, hasVisited, y, x)
                    else:
                        # It is ocean, mark as visited
                        hasVisited[y][x] = "0"

        return numOfIslands

    def searchEntireIsland(self, grid, hasVisited, y, x):
        if y >= 0 and x >= 0 and x < len(grid[0]) and y < len(grid):
            if grid[y][x] == "1" and hasVisited[y][x] == "-1":
                # Marking this spot on the island as 0
                hasVisited[y][x] = "0"

                # Need to search up down left and right now
                self.searchEntireIsland(grid, hasVisited, y + 1, x)
                self.searchEntireIsland(grid, hasVisited, y - 1, x)
                self.searchEntireIsland(grid, hasVisited, y, x + 1)
                self.searchEntireIsland(grid, hasVisited, y, x - 1)
        